crates/turborepo-lib/src/run/cache.rs

```diff
@@ -374,3 +382,88 @@ impl TaskCache {
         &self.expanded_outputs
     }
 }
+
+#[derive(Clone)]
+pub struct ConfigCache {
+    hash: String,
+    repo_root: AbsoluteSystemPathBuf,
+    config_file: AbsoluteSystemPathBuf,
+    anchored_path: AnchoredSystemPathBuf,
+    cache: Arc<AsyncCache>,
+}
+
+impl ConfigCache {
+    pub fn new(
+        hash: String,
+        repo_root: AbsoluteSystemPathBuf,
+        config_path: &[&str],
+        cache: Arc<AsyncCache>,
+    ) -> Self {
+        let config_file = repo_root.join_components(config_path);
+        ConfigCache {
+            hash,
+            repo_root: repo_root.clone(),
+            config_file: config_file.clone(),
+            anchored_path: AnchoredSystemPathBuf::relative_path_between(&repo_root, &config_file),
+            cache,
+        }
+    }
+
+    pub fn hash(&self) -> &str {
+        &self.hash
+    }
+
+    pub fn exists(&self) -> bool {
+        self.config_file.try_exists().unwrap_or(false)
+    }
+
+    pub async fn restore(
+        &self,
+    ) -> Result<Option<(CacheHitMetadata, Vec<AnchoredSystemPathBuf>)>, CacheError> {
+        self.cache.fetch(&self.repo_root, &self.hash).await
+    }
+
+    pub async fn save(&self) -> Result<(), CacheError> {
+        match self.exists() {
```

Comment:
There's a TOCTOU race condition where the file gets cleaned up between us checking for it's existence and it actually getting read into memory.

Reply:
Tracing through this, I don't think there's an easy way around it unless we rework the cache abstraction a fair bit. I think it's fine to leave as is 🤷